glom/matching.py

```diff
@@ -0,0 +1,241 @@
+"""
+Contains code associated with implementing match specs
+
+match mode tends to build expression trees by conjoining
+together matches with And / &, Or / |
+
+a "mode" provides new definitions for the meaning of
+basic python data structures inside the glom spec
+
+example use cases:
+
+filter list on condition (e.g. >= 10)
+[Coalesce(Match(int & (M >= 10)), default=SKIP)]
+
+ensure that dictionary values are all strings
+Match({DEFAULT: str})
+
+2 syntaxes for combining expressions:
+
+And(int, M > 0)
+M & int & (M > 0)
+
+"""
+from boltons.typeutils import make_sentinel
+
+import weakref
+
+from .core import GlomError, glom, T, Spec
+
+
+class GlomMatchError(GlomError): pass
```

Comment:
relatively big todo here: we need the path and so forth to craft actionable error messages.

Reply:
agree!

Reply:
short term: is there something to do to make `__repr__` kind of work as intended?  (without doing more upfront work than constructing a tuple)